src/error.rs

```diff
@@ -70,10 +78,17 @@ pub(crate) fn cert_result_to_error(result: rustls_result) -> rustls::Error {
     use rustls::Error::*;
     use rustls_result::*;
     match result {
-        CertInvalidEncoding => InvalidCertificateEncoding,
-        CertInvalidSignatureType => InvalidCertificateSignatureType,
-        CertInvalidSignature => InvalidCertificateSignature,
-        CertInvalidData => InvalidCertificateData("".into()),
+        CertEncodingBad => InvalidCertificate(CertificateError::BadEncoding),
+        CertExpired => InvalidCertificate(CertificateError::Expired),
+        CertNotYetValid => InvalidCertificate(CertificateError::NotValidYet),
+        CertRevoked => InvalidCertificate(CertificateError::Revoked),
+        CertUnhandledCriticalExtension => InvalidCertificate(CertificateError::UnhandledCriticalExtension),
+        CertUnknownIssuer => InvalidCertificate(CertificateError::UnknownIssuer),
+        CertBadSignature => InvalidCertificate(CertificateError::BadSignature),
+        CertNotValidForName => InvalidCertificate(CertificateError::NotValidForName),
+        CertInvalidPurpose => InvalidCertificate(CertificateError::InvalidPurpose),
+        CertApplicationVerificationFailure => InvalidCertificate(CertificateError::ApplicationVerificationFailure),
+        CertOtherError => InvalidCertificate(CertificateError::Other(Arc::<dyn std::error::Error>::new("".into()))), // TODO(cpu): busted!
```

Comment:
I was having a hard time figuring out the right way to contort an empty `&str` into a `Arc<dyn StdError + Send + Sync>`. I suspect I'm missing something obvious. This attempt doesn't compile :-)

Reply:
Not simple! There are two `From` impls that should get us there, and `Into` / `From` are supposed to apply recursively, but I think sometimes the language needs a little help:

https://doc.rust-lang.org/std/error/trait.Error.html#impl-From%3C%26str%3E-for-Box%3Cdyn%20Error%20+%20Send%20+%20Sync%20+%20%27a%2C%20Global%3E

https://doc.rust-lang.org/std/boxed/struct.Box.html#impl-From%3CBox%3CT%2C%20Global%3E%3E-for-Arc%3CT%3E

This ugly thing works for me:

```
        CertOtherError => InvalidCertificate(CertificateError::Other(
            Into::<Box<dyn std::error::Error + Send + Sync>>::into("").into(),
        )),
```

There may be a better way but I'm not finding it at the moment.

Reply:
Thank you! I was trying to `Box` in some of my other quick attempts at this but I was misinterpreting the problem and boxing the arc which didn't make a lot of sense (or solve the problem). It's comforting to know I was at least reaching in the right general direction.